src/workerd/util/entropy-source.h

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#pragma once
+
+#include <openssl/rand.h>
+
+#include <kj/common.h>
+#include <kj/compat/http.h>
+#include <kj/debug.h>
+
+#include <cstring>
+
+namespace workerd {
+
+// Buffered entropy source that maintains a thread-local pool of random bytes to reduce
+// RAND_bytes syscall overhead. This is especially beneficial on macOS where RAND_bytes
+// calls are expensive.
+//
+// The buffer is refilled in bulk (4KB at a time) when exhausted, amortizing syscall
+// overhead.
+class BufferedEntropySource final: public kj::EntropySource {
+ public:
+  BufferedEntropySource() = default;
+  ~BufferedEntropySource() noexcept(false) = default;
+  KJ_DISALLOW_COPY_AND_MOVE(BufferedEntropySource);
+
+  void generate(kj::ArrayPtr<kj::byte> buffer) override;
+
+  // Helper to generate multiple random 64-bit values in a single batch call to the entropy source.
+  // This significantly reduces overhead compared to multiple separate calls, especially for
+  // RAND_bytes on macOS where each call has substantial overhead.
+  template <size_t N>
+  static kj::Array<uint64_t> generateBatched(
```

Comment:
This is a rather large method. I'm not sure there's any real benefit in it being templated. It also likely doesn't really need to heap allocate the buffer. If you changed the signature to...

```c++
static void generateBatched(kj::ArrayPtr<uint64_t> out,
                            kj::Maybe<kj::EntropySource&> entropySource = kj::none);
```

... then you can move the body of the impl into the compilation unit and avoid the additional allocation on each call.